### PR TITLE
utils: add basic_xx_hasher

### PR DESCRIPTION
**Pull request by bhalevy**
Consolidate `bytes_view_hasher` and abstract_replication_strategy `factory_key_hasher` which are the same into a reusable utils::basic_xx_hasher.

To be used in a followup series for netw:msg_addr.

### CODE REVIEW

File: test/boost/hashers_test.cc [hunk 1]
```diff
@@ -102,3 +102,17 @@ SEASTAR_THREAD_TEST_CASE(mutation_fragment_sanity_check) {
         check_hash(f, 0x5092daca1b27ea26ull);
```
Comment (denesb) [inline test/boost/hashers_test.cc:102]: I think consistent header content is more important than avoiding a circular header dependency. Those are only a problem if they are cross-modules, but that is not the case here.

**Review by denesb — CHANGES_REQUESTED** [review]

File: test/boost/hashers_test.cc (hunk 1 above)
Comment (bhalevy) [reply test/boost/hashers_test.cc:102]: This means we'll need to implement this basic hasher in its own header that doesn't include bytes.hh.
I'll see if there's a better way around it...

**Comment by bhalevy** [comment]
In v2 ([62ea38d](https://github.com/scylladb/scylladb/commit/62ea38d9662e3991500cf41147e48c55119e68e3)):
- implemented simple_xx_hashser in a new header file: utils/simple_hashers.hh
- Defined BytesHasher and SimpleHasher concepts. (with respective static_asserts in utils/hashers.cc)

**Comment by denesb** [comment]
> * hh

I like this approach.
Why not name the new header `bytes_hasher.hh`, instead of `simple_hashers.hh` which is hard to differentiate from the existing `hashers.hh`.

**Review by denesb — APPROVED** [review]

File: utils/hashers.hh [hunk 2]
```diff
@@ -11,6 +11,12 @@
 #include "bytes.hh"
 #include "utils/hashing.hh"
 
+template<typename H>
+concept BytesHasher = Hasher<H> &&
+    requires (H& h) {
```
Comment (avikivity) [inline utils/hashers.hh:16]: BytesHasher evokes something that hashes bytes, not a hasher that returns bytes.

**Comment by bhalevy** [comment]
In v3 ([b3d91cb](https://github.com/scylladb/scylladb/commit/b3d91cbf653b56dc015780c9519fe263b374dccc)):
- Added a templated concept `HasherReturning`
  - base for `HasherReturningBytes` and `SimpleHasher`

**Review by denesb — APPROVED** [review]

**Pull request merged**